atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java

```diff
@@ -336,11 +345,82 @@ public void hostIsAutomaticallyAddedOnRefresh() {
+    }
+
+    @Test
+    public void onlyValidatedHostsAreAddedOnRefresh() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+
+        setCassandraServersTo(CASS_SERVER_1);
+
+        createClientPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1);
+        setupHostsWithInconsistentTopology(CASS_SERVER_2);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+        refreshPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1, CASS_SERVER_3);
+        assertThat(absentHostTracker.returnPool(CASS_SERVER_2)).isPresent();
+    }
+
+    @Test
+    public void hostsWithDifferingTopologiesTrackedAsAbsentHost() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2, CASS_SERVER_3);
+        setupHostsWithInconsistentTopology(CASS_SERVER_2);
+        createClientPool();
+        assertThat(absentHostTracker.returnPool(CASS_SERVER_2)).isPresent();
+    }
+
+    @Test
+    public void throwsWhenNoServersInPoolAndServersAddedPresentButAreInvalid() {
+        setupHostsWithInconsistentTopology(CASS_SERVER_1);
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+        setCassandraServersTo(CASS_SERVER_1);
+        assertThatThrownBy(this::createClientPool).isInstanceOf(SafeIllegalStateException.class);
+    }
+
+    @Test
+    public void reuseAbsentHostsContainerIfPresent() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy(), CASS_SERVER_2.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
+        createClientPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1, CASS_SERVER_2);
+        CassandraClientPoolingContainer cassandraServerTwoContainer = poolServers.get(CASS_SERVER_2);
+
+        setCassandraServersTo(CASS_SERVER_1);
+        refreshPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
+        refreshPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1, CASS_SERVER_2);
+        assertThat(cassandraServerTwoContainer).isEqualTo(poolServers.get(CASS_SERVER_2));
+    }
+
+    @Test
+    public void noServersAddedOrRemovedWhenCassandraServerListStatic() {
+        setupThriftServers(ImmutableSet.of(CASS_SERVER_1.proxy()));
+        when(config.autoRefreshNodes()).thenReturn(true);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
+
+        createClientPool();
+        assertThatMap(poolServers).containsOnlyKeys(CASS_SERVER_1, CASS_SERVER_2);
+
+        setCassandraServersTo(CASS_SERVER_1, CASS_SERVER_2);
```

Comment:
nit: I don't personally think this line adds much (at least, it seems to me if you don't call set cassandra servers, the list is static). This is a minor point if you feel it helps readability.

Reply:
let's remove, my aim here was for readability, but as you said yourself you don't think it adds much :)